src/worker/SandboxedWorker.js

```diff
@@ -0,0 +1,146 @@
+ * script that manipulated some indexedDB databases.
+ *
+ * One way to solve this problem is to use a data URL instead of a string or
+ * object URL. WebWorkers created with data URLs are created with opaque
+ * ("orphaned") origins (see https://html.spec.whatwg.org/multipage/workers.html#dom-worker)
+ * and this negates the same-origin access.
+ *
+ * However, consuming and processing data URLs was found to be hugely
+ * inefficient and resulted in severe performance degradation upon the initial
+ * load of a script.
+ *
+ * Thus, we instead chose to "wrap" these WebWorkers with an iframe, using the
+ * iframe to create the opaque origin.
+ */
+
+class SandboxedWorker extends EventTarget {
+  constructor(scriptUrl, { name } = {}) {
+    super()
+
+    this.name = name
+    this.iframe = document.createElement('iframe')
+    this.iframe.sandbox = 'allow-scripts'
+    this.iframe.style = 'position: absolute; width: 0; height: 0; opacity:0;'
+
+    const source = `
+      <script>
+        const fetchScriptUrlAsBlob = async url => {
+          // In the future, we might support IPFS protocols in addition to http
+          const res = await fetch(url, {
+            method: 'GET',
+            mode: 'cors',
+          })
+          // If status is not a 2xx (based on Response.ok), assume it's an error
+          // See https://developer.mozilla.org/en-US/docs/Web/API/GlobalFetch/fetch
+          if (!(res && res.ok)) {
+            throw res
+          }
+          return res.blob()
+        }
+        async function getObjectUrlForScript(scriptUrl) {
+          const blob = await fetchScriptUrlAsBlob(scriptUrl)
+          return URL.createObjectURL(blob)
+        }
+        const init = async () => {
+          // WebWorkers can only load scripts from the local origin, so we
+          // have to fetch the script (from an IPFS gateway) and process it locally.
+          const workerUrl = await getObjectUrlForScript('${scriptUrl}')
+
+          const worker = new Worker(workerUrl, { name: '${name}' })
+
+          // Must use '*' for origin as we've sandboxed the iframe's origin
+          worker.addEventListener('message', event => window.parent.postMessage({ from: '${name}', msg: event.data }, '*'), false)
+          worker.addEventListener(
+            'error',
+            error => {
+              console.error('Error from worker for ${name} (loaded from ${scriptUrl}):', error.message, error)
+              window.parent.postMessage(
+                {
+                  from: '${name}',
+                  error: {
+                    filename: error.filename,
+                    message: error.message,
+                    lineno: error.lineno,
+                  },
+                },
+                '*'
+              )
+            },
+            false
+            )
+          window.addEventListener('message', ({ data, source }) => {
+            if (source === window.parent) {
```

Comment:
This is also important; we need to check the source of the message, otherwise it could be another window sending information to this iframe.